_posts/2020-10-04-quarkus-native-on-aarch64.adoc

```diff
@@ -0,0 +1,408 @@
+
+## Prerequisites
+
+* *Single-board computer:* The ARM-compatible CPU integrated into the single-board computer must support 64-bit mode. 
+
+* *Operating system:* 
+** To make a Quarkus native container image run in a container, a 64-bit Linux OS is required;
+
+** The OS must support aarch64 CPUs, which leads us to the need for an aarch64-base distribution available for that OS;
+
+** Nice to have: the target OS should fully support the target single-board computer.**
+
+* *Container technology:* surprise, surprise! an OCI standard container engine is compulsory to run containers on Linux :-)
+
+* *Build tools:* as per the large amount of resources required by the GraalVM to compile Quarkus in native mode, an additional we need an ARM server / Virtual machine to achieve that goal. The server should meet the following minimum requirements: 
+** CPU: same ARM model as the one embedded in the target single-board computer
+** Cores: 4
+** Ram: 8Gb
+** Disk space: 20Gb
+
+* *Container registry:* In order to be able to deploy onto the single-board computer the container image containing the Quarkus native app built in the ARM server, an account on a remote container registry is required.
+
+
+
+
+
+## Technology
+
+### Single-sensor board
+
+Choosing the target single-board computer model has been quite an easy task: I already had a Raspberry Pi meeting the requirements above:
+
+* _Model:_ Raspberry Pi 3 B+
+* _SOC:_ Broadcom BCM2837B0, Cortex-A53 (ARMv8) 64-bit SoC
+* _CPU:_ 1.4GHz 64-bit quad-core ARM Cortex-A53 CPU
+* _RAM:_ 1GB LPDDR2 SDRAM
+* _WIFI:_ Dual-band 802.11ac wireless LAN (2.4GHz and 5GHz ) and Bluetooth 4.2
+* _Ethernet:_ Gigabit Ethernet over USB 2.0 (max 300 Mbps). Power-over-Ethernet support (with separate PoE HAT). Improved PXE network and USB mass-storage booting.
+* _Thermal management:_ Yes
+* _Video:_ Yes – VideoCore IV 3D. Full-size HDMI
+* _Audio:_ Yes
+* _USB 2.0:_ 4 ports
+* _GPIO:_ 40-pin
+* _Power:_ 5V/2.5A DC power input
+* _Operating system support:_ Linux and Unix
+
+### OS
+
+I’ve tried a few operating system meeting the requirements listed above. The two possible solutions were Fedora Internet of Things Edition and Gentoo.
+After some tests, I’ve picked up https://iot.fedoraproject.org/[Fedora IoT] for the following reasons:
+
+* Full, native support to 64-bit architecture (aarch64)
+* Stability
+* Usability
+* Maturity
+* Lifecycle
+* Full support on RPi 3 B+
+
+### Container technology
+
+This one has been an easy task as well: https://podman.io/[Podman]. Podman is a daemonless container engine for developing, managing, and running OCI Containers on your Linux System. Containers can either be run as root or in rootless mode. Moreover, there’s no need for installing additional components/layers to Fedora IoT basic installation as the Podman container engine is already available in the OS distribution.
```

Comment:
this is probably unrelated but how are you overcoming the fact cgroups on fedora is now v2 so almost any docker based setup I try gets warning and eventually fail ? does this require using older fedora or is this just not a concern for this case ?

Reply:
That's a good point. I haven't faced any issue with groups running Podman containers on both Fedora IoT 31 and 32.

That is explained in the following article: https://podman.io/blogs/2019/10/29/podman-crun-f31.html
I'm quoting the quick explanation:
_"The crun command is the runtime the allows for cgroup V2 support and is supplied starting with Fedora 31. Other container systems use the runc runtime. However, runc only supports cgroup V1. The cgroup kernel feature allows you to allocate resources such as CPU time, network bandwidth and system memory to a container. Version 1 of cgroup only supports containers that are run by root, while version 2 supports containers that are run by root or a non-privileged user."_
The output from the command "podman info" on the RPi related to the ico runtime is the following:
```
ociRuntime:
    name: crun
    package: crun-0.14.1-4.fc32.aarch64
    path: /usr/bin/crun
    version: |-
      crun version 0.14.1
      commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
```
Meaning the Podman version already included in Fedora IoT 31 and later fully supports cgroup V2 (if I'm not mistaken)

Reply:
Great. My struggles probably been using stuff that expected some root.